Fix 2:1 balance on artificial cells over periodic boundaries.

kronbichler opened the pull request:
Currently, the implementation copies parts of the implementation from source/grid/tria.cc, `fix_coarsen_flags` method (but inserting an indirection that connects periodic vertices).

A general question to discuss is regarding the support for periodic boundary conditions. This PR opts for merely correcting the functionality for multigrid of `parallel::distributed::Triangulation` that already supports `add_periodicity` and the corresponding p4est functionality.

An alternative would be to make periodic boundaries first-class citizens in `dealii::Triangulation` with correct mesh balance over the periodic boundary, which is a lot of work, though. In that context, also something like `cell->neighbor()` should be able to look beyond the periodic boundary, rather than returning `at_boundary()`. Opinions?

tjhei commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,204 @@ namespace parallel
+      identify_periodic_vertices_recursively(const GridTools::PeriodicFacePair<ITERATOR> &periodic,
+                                             std::vector<unsigned int> &topological_vertex_numbering)
+      {
+        // for hanging nodes we will consider all necessary coupling already
```
there are three cases:
1. both have children: you recurse, but you also drop out of the if and identify vertices (is that necessary?)
2. both have no children: you identify the vertices (okay)
3. one of them has children: here you end up identifying the wrong vertices, I think.

kronbichler replied:
Case 3 should be fine because on adaptive meshes there is no corresponding other vertex id of hanging nodes. The vertices which appear on both sides are identified on the next coarser level.

tjhei commented on source/distributed/tria.cc:
```diff
@@ -3272,6 +3470,11 @@ namespace parallel
 
+          // enforce 2:1 mesh balance over periodic boundaries
+          if (this->smooth_grid & dealii::Triangulation<dim,spacedim>::limit_level_difference_at_vertices)
+            enforce_mesh_balance_over_periodic_boundaries(*this,
```
Are we sure this produces the correct result with a single iteration? Can you add a check (maybe wrapped in ``#ifdef DEBUG``) that runs ``enforce_mesh_balance_over_periodic_boundaries`` after ``execute_coarsening_and_refinement`` below to check if any cells will be flagged?

bangerth replied:
I think this is the wrong place. We have always advertised `prepare_coarsening_and_refinement()` as the function that definitively determines which cells will be coarsened and which will be refined. We need this, for example, for `SolutionTransfer`. 

The right approach is likely to make `prepare_c_and_r()` virtual, and override it in the current class; there, you'd iterate between `::Triangulation::prepare_c_and_r()` and `enforce_mesh_balance_over_boundaries()` until the refine/coarsening flags don't change any more (in much the same way as `::Triangulation::prepare_c_and_r()` already iterates). Can you rewrite the patch in this direction?

kronbichler replied:
Regarding the single iteration: It can require more iterations, but there is already an iteration `while(mesh_changed)` around that code block in the distributed triangulation.

Regarding `prepare_coarsening_and_refinement`: I have rewritten the patch that way. The overridden `prepare_coarsening_and_refinement` method calls the implementation of the base class and `enforce_mesh_balance_over_periodic_boundaries`, looping until nothing changes any more. Without periodic boundaries (or no `limit_level_difference_at_vertices` which is only set when the multigrid hierarchy is requested), the method is called once. Otherwise, we will ensure that the two methods have 'converged'.

tjhei commented:
(I am sitting next to Martin, so it would be good if somebody else also takes a look at this)

> Opinions?

Ideally periodicity support ends up inside ``Triangulation`` but I agree that this is a ton of work. My vote would be to merge this for now to fix geometric multigrid.

bangerth commented:
It's likely going to be a lot of work to make this happen inside `Triangulation`, simply because you would somehow have to deal with the fact that currently `cell->face(f) != cell->neighbor(f)->face(neighbor_of_neighbor(f))` even if you manage to get the neighbor pointers right. The reason is that faces have vertex indices and vertex indices point to coordinates, and so the face that connects periodic faces happens to be two faces.

It's been said many times that that may well have been one of the original sins of deal.II. Quite possibly so, but rectifying this in a way that's halfway backward compatible is just going to be hard, so I agree with the current approach.

There is also an issue on which I'm not entirely clear. It is conceivable that someone may want to solve a problem in which one of the variables is periodic and the other is not. I have no example for this, but at least currently that's possible. You can only do that if periodic boundaries are not first class citizens of `Triangulation`, but are in fact only applied at the degree of freedom level, as they are currently. There is virtue to this, and the fact that you just happen to fix up the 2:1 balance relationship doesn't get in the way of it.

kronbichler commented:
OK, we seem to agree that we choose a small solution now. I think I have addressed all questions (and fixed a bug that first appeared on 448 MPI ranks together with @tjhei).

tjhei commented on tests/mpi/distribute_mg_dofs_periodicity_01.mpirun=28.output:
```diff
@@ -0,0 +1,7 @@
+
+DEAL::Number of cells: 1024
```
is there much value in this test with 28 procs now that we fixed things?

tjhei replied:
(worrying about cost of running this)

kronbichler replied:
In 3D, the problem only failed on 28 processors (the 6 processor case was for 2D). Thus, I think that it represents a case that previously failed but now runs. We could argue that the underlying problem appears in 2D already, so there is no real need to keep it.

kronbichler replied:
The run time on my notebook (2 cores + hyperthreading) is ~60s in debug mode on 28 MPI processes, as compared to ~30s on 6 processes. For a test machine with 10+ core I don't think the difference will be too large.

bangerth replied:
That would certainly make it one of the longer running tests, but I think I'd be ok with it if you think that it adds something useful. On my 64 core machines, a testsuite run takes ~35 minutes. This might add 5-10 seconds to it, which is a lot for one test, but then we also don't add these kinds of tests very often.

kronbichler replied:
I think it is good to keep one 3d test where it is a bit more difficult to guess the level ghost layer.

tjhei commented:
Looks good to me! :+1:

bangerth commented:
On 01/27/2016 11:43 AM, Martin Kronbichler wrote:
> (and fixed a bug that first appeared on 448 MPI ranks together with @tjhei
> <https://github.com/tjhei>).

Impressive. How did you even debug this?

kronbichler commented:
> > (and fixed a bug that first appeared on 448 MPI ranks together with @tjhei

> Impressive. How did you even debug this?

We were thinking a while for exactly how you go about this kind of problem (a 64^3 mesh would probably fit into memory for a single machine, so you could probably add some output at the right spot to identify the bad processor), but thought is was worth to go through the case per hand on a single cell in 3d with periodic boundaries. There we could see that the identification was wrong - I had mixed the "topological_vertex_index" with the original ones in a bad way. :-)

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,237 @@ namespace parallel
+      while ((this->smooth_grid & dealii::Triangulation<dim,spacedim>::limit_level_difference_at_vertices)
+             &&
+             enforce_mesh_balance_over_periodic_boundaries(*this,
+                                                           periodic_face_pairs_level_0));
```
I find this difficult to read. The verb `enforce` to me indicates that the function *does* something, not that it computes a condition that can be tested. Maybe write it as
```
  bool mesh_changed = false;   // or some other name
  do
    {
        this->Base::prepare_c_and_r();
        if (smooth_grid & ...)
          mesh_changed = enforce_mesh_...()
      }
  while (mesh_changed);
```
?

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,237 @@ namespace parallel
+      template <typename ITERATOR>
+      void
+      identify_periodic_vertices_recursively(const GridTools::PeriodicFacePair<ITERATOR> &periodic,
+                                             std::vector<unsigned int> &topological_vertex_numbering)
```
I don't think I quite understand what `topologic_vertex_numbering` is supposed to be. Can you add a comment to the function?

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,237 @@ namespace parallel
+                  {
+                    if (periodic.cell[0]->child(c)->face(periodic.face_idx[0]) ==
+                        periodic.cell[0]->face(periodic.face_idx[0])->child(cf))
+                      break;
```
You are looking for `GeometryInfo::child_cell_on_face` :-)

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,237 @@ namespace parallel
+                for (; c<periodic.cell[1]->n_children(); ++c)
+                  {
+                    if (periodic.cell[1]->child(c)->face(periodic.face_idx[1]) ==
+                        periodic.cell[1]->face(periodic.face_idx[1])->child(cf))
```
same here

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,237 @@ namespace parallel
+        // TODO: fix non-standard orientation
+        Assert(periodic.orientation[0] == true &&
+               periodic.orientation[1] == false &&
+               periodic.orientation[2] == false,
```
This should also be addressed by using the function in `GeometryInfo` mentioned above.

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,237 @@ namespace parallel
+            const unsigned int vi0 = topological_vertex_numbering[periodic.cell[0]->face(periodic.face_idx[0])->vertex_index(v)];
+            const unsigned int vi1 = topological_vertex_numbering[periodic.cell[1]->face(periodic.face_idx[1])->vertex_index(v)];
+            const unsigned int min_index = std::min(vi0, vi1);
+            topological_vertex_numbering[periodic.cell[0]->face(periodic.face_idx[0])->vertex_index(v)] = topological_vertex_numbering[periodic.cell[1]->face(periodic.face_idx[1])->vertex_index(v)] = min_index;
```
That's a rather long line ;-)

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3122,6 +3122,237 @@ namespace parallel
+      template <int dim, int spacedim>
+      bool enforce_mesh_balance_over_periodic_boundaries
+      (Triangulation<dim,spacedim> &tria,
+       std::vector<GridTools::PeriodicFacePair<typename dealii::Triangulation<dim,spacedim>::cell_iterator> > periodic_face_pairs_level_0)
```
take the second argument by const reference instead?

kronbichler commented:
Thanks for the pointer regarding `GeometryInfo`. That helped me fix the non-standard orientation case, too. I think I have addressed all comments now.
